vta/hardware/chisel/src/main/scala/core/TensorAlu.scala

```diff
@@ -0,0 +1,316 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package vta.core
+
+import chisel3._
+import chisel3.util._
+import vta.util.config._
+
+class Alu(implicit p: Parameters) extends Module {
+  val aluBits = p(CoreKey).accBits
+  val io = IO(new Bundle {
+    val opcode = Input(UInt(C_ALU_OP_BITS.W))
+    val a = Input(SInt(aluBits.W))
+    val b = Input(SInt(aluBits.W))
+    val y = Output(SInt(aluBits.W))
+  })
+  val min = Mux(io.a < io.b, io.a, io.b)
+  val max = Mux(io.a < io.b, io.b, io.a)
+  val add = io.a + io.b
+
+  val ub = io.b.asUInt
+  val width = log2Ceil(aluBits)
+  val m = ~ub(width - 1, 0) + 1.U // FIXME: there should be a SHL and SHR instruction
+  val n = ub(width - 1, 0)
+  val shl = io.a << m
+  val shr = io.a >> n
+
+  io.y :=
+    MuxLookup(io.opcode,
+              io.a, // default
+      Array(
```

Comment:
Not something we need to handle now, but maybe we could make this mux/instruction/etc bit extensible so we can add new ops and turn on/off without needing to modify the core ALU logic?

Reply:
Alright, [this](https://github.com/vegaluisjose/tvm/commit/7b3f2de8d9734513279660d4a4d083b847f765f9) is the best I could do for making ALU a bit more programmable.